recipes/lora_finetune_distributed.py

```diff
@@ -40,28 +40,57 @@
     """
-    Distributed LoRA finetuning recipe for dense transformer-based LLMs such as Llama2.
+    Distributed LoRA finetuning recipe for dense transformer-based LLMs such as Llama2. This recipe supports
+    distributed training and can be run on a single node (1 to 8 GPUs).
 
-    This recipe supports:
-        - FSDP and activation checkpointing. Activation Checkpointing is configurable and is disabled
-            by default.
-        - Full fp32 and full bf16 training. Mixed precision training or fp16 are currently not
+    Features:
+        - FSDP. Supported using PyTorch's FSDP APIs. DDP is currently not supported. Traning on CPU is not
             supported.
-        - Checkpointing of full model weights and optionally of optimizer state (for
-            checkpoints created during training).
-        - Logging to terminal, WandB, or TensorBoard.
-
-    Assumptions:
-        - Training happens on CUDA (CPU training is not supported)
-        - Checkpoints are ONLY saved at epoch boundaries. In case of failure, work done
-            in ongoing epoch is lost.
-        - Datasets are Map-style and data fits in memory (not streamed).
-
-    The following configs can be used to run this recipe:
-        >>> tune ls
-        RECIPE                         CONFIG
-        lora_finetune_distributed      llama2/7B_lora
-                                       llama2/13B_lora
+
+        - Activation Checkpointing. This can be controlled using the ``activation_checkpointing``
+            flag. Activation checkpointing helps reduce the memory footprint since we no longer keep
+            activations in memory and recompute them during the backward pass. This is especially helpful
+            for larger batch sizes when you're memory constrained. But these savings in memory come at the
+            cost of training performance. In most cases training can slow-down quite a bit as a result of
+            this activation recomputation.
+
+        - Precision. Full fp32 and bf16 training are supported. Precision is controlled using the ``dtype``
+            flag. When ``dtype=bf16``, all activations, gradients and optimizer states are in bfloat16. In
+            most cases this should halve the memory footprint of full precision (fp32) training, without
+            loss in model quality (will depend on the model, training data and other settings). For
+            GPUs which do not support bfloat16, we fall back to fp32. Mixed precision training and fp16
+            precision are currently not supported.
+
+        - Gradient Accumulation. You can simulate larger batch sizes by accumulating gradients. This is
+            controlled using the ``gradient_accumulation_steps`` flag.
+
+                Total Batch Size = batch_size * number of GPUs * gradient accumulation steps.
+
+            For example: with batch_size=1, nproc_per_node=2 and gradient_accumulation_steps=32 we get a
+            total batch size of 64.
```

Comment:
Idk if this is the right place to point this out or not, but there are also implications for learning rate when tuning this value

Reply:
Do you mean LR should be appropriately tuned? Hmm yeh its a good point, but not sure if I want to point out here. I do want to capture how to think about BS when using gradient accumulation though. let me know if you feel strongly about the LR